app/src/main/java/com/lambdasoup/quickfit/persist/SyncWorker.kt

```diff
@@ -0,0 +1,156 @@
+                                            insertNextSession(sessionsCursor, completer, SyncStatus(hasFailedInsertions = false))
+                                        } catch (e: RemoteException) {
+                                            completer.set(Result.retry())
+                                        }
+                                    }
+
+                                    override fun onConnectionSuspended(i: Int) {
+                                        Timber.d("connection suspended")
+                                        if (i == GoogleApiClient.ConnectionCallbacks.CAUSE_NETWORK_LOST) {
+                                            Timber.d("Connection lost.  Cause: Network Lost.")
+                                        } else if (i == GoogleApiClient.ConnectionCallbacks.CAUSE_SERVICE_DISCONNECTED) {
+                                            Timber.d("Connection lost.  Reason: Service Disconnected")
+                                        }
+                                        completer.set(Result.retry())
+                                    }
+                                }
+                        )
+                        .addOnConnectionFailedListener { result ->
+                            Timber.d("connection failed: %s", result.errorMessage)
+                            appContext.startService(FitApiFailureResolutionService.getFailureResolutionIntent(appContext, result))
+                            completer.set(Result.failure())
+                        }
+                        .build()
+
+                completer.addCancellationListener(Runnable { googleApiClient?.disconnect() }, Executors.newSingleThreadExecutor())
+
+                Timber.d("starting sync work")
+                googleApiClient!!.connect()
+            }
+
+    private fun insertNextSession(cursor: Cursor?, completer: CallbackToFutureAdapter.Completer<Result>, syncStatus: SyncStatus) {
+        if (cursor == null || !cursor.moveToNext()) {
+            // done with sessions
+            Timber.d("Done; disconnecting.")
+            googleApiClient!!.disconnect()
+            // sync finished
+            completer.set(if (syncStatus.hasFailedInsertions) Result.retry() else Result.success())
+            return
+        }
+
+        val startTime = cursor.getLong(cursor.getColumnIndex(QuickFitContract.SessionEntry.START_TIME))
+        val endTime = cursor.getLong(cursor.getColumnIndex(QuickFitContract.SessionEntry.END_TIME))
+        val sessionName = cursor.getString(cursor.getColumnIndex(QuickFitContract.SessionEntry.NAME))
+        val sessionBuilder = Session.Builder()
+                .setActivity(cursor.getString(cursor.getColumnIndex(QuickFitContract.SessionEntry.ACTIVITY_TYPE)))
+                .setStartTime(startTime, TimeUnit.MILLISECONDS)
+                .setEndTime(endTime, TimeUnit.MILLISECONDS)
+        if (sessionName != null) {
+            sessionBuilder.setName(sessionName)
+        }
```

Comment:
`else`?

Reply:
..else we don't set a name for the session. That's fine.

Could do some abuse of kotlin here, like
```
                .setEndTime(endTime, TimeUnit.MILLISECONDS)
                .apply { 
                    sessionName?.let { setName(it) }
                }
```

but I see no advantage there. In fact, this relies on the fact that the Builder will mutate _and_ return itself.

Reply:
I did not know that not having a name was ok. Just made sure you give this a second thought. Maybe `#setName(String)` can also be called with `null`, then you would not need the `if` in the first place.